destination.go

```diff
@@ -69,24 +51,119 @@ func (d *Destination) Configure(ctx context.Context, cfg config.Config) error {
 	return nil
 }
 
-func (d *Destination) Open(_ context.Context) error {
-	// Open is called after Configure to signal the plugin it can prepare to
-	// start writing records. If needed, the plugin should open connections in
-	// this function.
+func (d *Destination) Open(_ context.Context) (err error) {
+	d.db, err = sqlx.Open("mysql", d.config.URL)
+	if err != nil {
+		return fmt.Errorf("failed to connect to mysql: %w", err)
+	}
+
 	return nil
 }
 
-func (d *Destination) Write(_ context.Context, _ []opencdc.Record) (int, error) {
-	// Write writes len(r) records from r to the destination right away without
-	// caching. It should return the number of records written from r
-	// (0 <= n <= len(r)) and any error encountered that caused the write to
-	// stop early. Write must return a non-nil error if it returns n < len(r).
-	return 0, nil
+func (d *Destination) Write(ctx context.Context, recs []opencdc.Record) (int, error) {
+	for _, rec := range recs {
+		switch rec.Operation {
+		case opencdc.OperationSnapshot:
+			if err := d.upsertRecord(ctx, rec); err != nil {
+				return 0, err
+			}
+		case opencdc.OperationCreate:
+			if err := d.upsertRecord(ctx, rec); err != nil {
+				return 0, err
+			}
+		case opencdc.OperationUpdate:
+			if err := d.upsertRecord(ctx, rec); err != nil {
+				return 0, err
+			}
+		case opencdc.OperationDelete:
+			if err := d.deleteRecord(ctx, rec); err != nil {
+				return 0, err
+			}
+		}
+	}
+
+	return len(recs), nil
 }
 
 func (d *Destination) Teardown(_ context.Context) error {
-	// Teardown signals to the plugin that all records were written and there
-	// will be no more calls to any other function. After Teardown returns, the
-	// plugin should be ready for a graceful shutdown.
+	if d.db != nil {
+		if err := d.db.Close(); err != nil {
+			return fmt.Errorf("failed to close connection: %w", err)
+		}
+	}
+
 	return nil
 }
+
+func (d *Destination) upsertRecord(ctx context.Context, rec opencdc.Record) error {
+	payload, isStructured := rec.Payload.After.(opencdc.StructuredData)
+	if !isStructured {
+		data := make(opencdc.StructuredData)
+		if err := json.Unmarshal(rec.Payload.After.Bytes(), &data); err != nil {
+			return fmt.Errorf("failed to json unmarshal non structured data: %w", err)
+		}
+
+		payload = data
+	}
+
+	columns := make([]string, 0, len(payload))
+	placeholders := make([]string, 0, len(payload))
+	values := make([]any, 0, len(payload))
+	upsertList := make([]string, 0, len(payload))
+
+	for col, val := range payload {
+		columns = append(columns, col)
+		placeholders = append(placeholders, "?")
+		values = append(values, val)
+
+		upsertList = append(upsertList, fmt.Sprint(col, " = VALUES(", col, ")"))
+	}
+
+	query := fmt.Sprintf(`
```

Comment:
I'd like to suggest using an SQL query builder, that will help us with building a query, sanitizing values, etc.

Reply:
I tried to use squirrel, but frankly, I got more boilerplate than the version without an sql builder. The code already uses parametrized queries, so it should be safe from sql injections.

I tried to port the code to use github.com/Masterminds/squirrel, but it results in more boilerplate.

Reply:
Well let me see, it would be consistent with the postgres connector.